### ISSUE TITLE
All commands giving WARNING see url for more info. 

### ISSUE THREAD

**Issue by NealWalters** [ngetchell/PSGitLab#174]
## 1. Provide a general summary of the issue in the Title above
First time user, trying any command, get response: 
WARNING: . See https://git.mydomain.com/help/api/README.md#status-codes for more information.

## 2. Describe Your Environment
Operating System and PowerShell version:
Win 2012/R2 Standard; Powershell 4.0; then  upgraded to Powershell 5.0.  Same issue. 

## 3. Expected Behavior
I'm a new user, but expect to get some data back, not the warning above. \

## 4.Current Behavior
This is what I'm running: 

Save-GitLabAPIConfiguration -Token 'xxxx' -Domain 'https://git.mydomain.com' -APIVersion 3
Get-GitLabUser -Username "myuser"
Result: 
WARNING: . See https://git.mydomain.com/help/api/README.md#status-codes for more information.

## 5. Possible Solution
No idea. 

## 6. Context
First time user; I could very well be doing something wrong. 
Thanks

**Comment by ngetchell** [comment]
What version of GitLab are you using. Going to see if I can reproduce the issue to debug.

**Comment by NealWalters** [comment]
GitLab Community Edition 10.3.3 - I think probably I'm just doing something wrong, but don't know what.

**Comment by NealWalters** [comment]
I added a Get-GitLabProject  after my Get-GitLabUser command. 
I had some more time to "play" today, and Get-GitLabProject works  works when I use http instead of https.  If I do the Get-GitLabProject with https I see "WARNING: . See https://git.echo.com/help/api/README.md#status-codes for more information."

My Get-GitLabUser is still failing.  If I take off the "-Username" parameter (and use http) I now get a list of all users.  Maybe it stops after 20 or so users; but I don't see my username on the list returned.  Do I have to call it in a loop to get more?  I tried "Get-GitLabUser -All" and same result set. 

Maybe when you run 'Get-GitLabUser -Username "baduser"' it is getting a 404 not found, and returning "WARNING: . See https://git.echo.com/help/api/README.md#status-codes for more information." and not the specific status code? 

Neal

**Comment by robinmalik** [comment]
With `Get-GitlabUser -Username validuser` I get:
"WARNING: Not Found. See https://gitlab.lboro.ac.uk/help/api/README.md#status-codes for more information."

but also, a valid returned user object. Same result with API v3 and v4.

Running with `-Verbose` seems to suggest it's a pagination issue. As soon as a query is made to (example) `https://gitlab.domain.com/api/v4/users&page=2` it returns a 404. I think the problem is that in the URL where there is an `&`, there should be a `?`.

**Comment by DHowett** [comment]
The `https` issue sounds suspiciously like #146. Do you have a build of PSGitLab from after #171 was merged (~6 days ago)?

**Comment by robinmalik** [comment]
I can confirm that the version I'm running uses the `[Net.ServicePointManager]::SecurityProtocol += [Net.SecurityProtocolType]::Tls12` fix.

I did get the same error as [this](https://github.com/ngetchell/PSGitLab/issues/146#issuecomment-369343543) when running `Test-GitLabAPI` (i.e. `$Result = Invoke-WebRequest -UseBasicParsing -Uri "$Domain/api/$Ver...`) **but** after running another command, Test-GitLabAPI worked as you'd expect. The problem with this command specifically, is that it doesn't use `QueryGitLabAPI` which is where the TLS fix is set for the session. I'll raise this as a new issue.

I think perhaps the issue I have with `Get-GitLabUser` is slightly different to Neal (and he may benefit from the later build). For example, `Get-GitLabProject` works perfectly fine for me with or without `https`. 

For myself, with `Get-GitLabUser` I get the warning (not found) when running with `All`, `Username` and `Email` - basically whenever it needs to grab all records beyond the default pagination size of 20. Verbose shows that it's executing `https://gitlab.domain.com/api/v4/users&page=2` and this will never work. I need to dig into it a little deeper as I'm unsure why pagination would work for other functions but not this one.

Edit: Could be that pagination is being fixed via #168 and/or #169

**Comment by NealWalters** [comment]
Not sure if this is how to get latest, I re-ran: 
Find-Module -Name PSGitLab | Install-Module

then to confirm did: 
Get-Module PSGitLab 

ModuleType Version    Name                                ExportedCommands                                                                               
---------- -------    ----                                ----------------                                                                               
Script     2.7.0      PSGitLab                            {Add-GitLabProjectRunner, Block-GitLabUser, Close-GitLabMergeRequest, Close-GitLabMilestone...}

But I'm not sure where to xref the issue fixes to the Version #. 

I can now run Get-GitLabProject with http instead of https. 

So even though the issue was fixed, should the error originally reported have been more clear? 
My original error: 
WARNING: . See https://git.mydomain.com/help/api/README.md#status-codes for more information.

was quite meaningless, and the link didn't even work (The Site can't be reached). 
Seems like it should display the status code then tell me that I got a bad status code. 

Thanks for the help!  I think my main issue is basically a duplicate of #146. 

The username issue might be for another issue, but that was just an example command that I picked.

**Comment by NealWalters** [comment]
This is occurring again, not sure why I closed it originally.  These two issues #177 and #174 seem to be interlinked, fix one and the other breaks or vise-versa. 
https://github.com/ngetchell/PSGitLab/issues/177

Could I get a small sample to just get a project and show the branches or commits? 

Thanks,
Neal

**Comment by NealWalters** [comment]
Changing the https to http (on the Save-GitLabAPIConfiguration statement) seems to help get some variables back. 
Is that still an issue? 

Thanks,
Neal

**Comment by NealWalters** [comment]
Seems like I can run one or two commands (with http instead of https), then I start getting the "You cannot call a method on a null-valued expression. 
Neal

**Comment by NealWalters** [comment]
Here's exactly what I'm running except masked names.  Seems like once it runs a command, then the Powershell environment is corrupt, and I have to close it and re-open it. 

```
cls
$PSVersionTable.PSVersion 

#NOTE: Have had problems using https instead of http in the line below. 
Write-Host "Save-GitLabAPIConfiguration " 
Save-GitLabAPIConfiguration -Token 'xxxxx' -Domain 'http://git.xxxxxxxxx.com' -APIVersion 3
 
Write-Host " "
Write-Host "$project=Get-GitLabProject -Search 'MyProject'" 
$project = Get-GitLabProject -Search "MyProject" 

#Get all branches for above project 
Write-Host "Get-GitLabProjectBranch -ProjectID $project.id  (try to get all branches on prior project) " 
Get-GitLabProjectBranch -ProjectID $project.id 

$showDate = Get-Date -DisplayHint Date
Write-Host "Completed at $showDate"

```

Output from first time run: 


```
Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      14409  1005    
Save-GitLabAPIConfiguration 
WARNING: All tokens will be sent over HTTP. Recommendation: Use HTTPS.
 
=Get-GitLabProject -Search 'MyProject'
Get-GitLabProjectBranch -ProjectID @{id=577; description=; default_branch=master; tag_list=System.Object[]; public=False; archived=False; visibility_level=10; ssh_u
rl_to_repo=git@git.xxxxxxxx.com:Echo.BSSR/MyProject.git; http_url_to_repo=http://git.xxxxxxxx.com/Echo.BSSR/MyProject.git; web
_url=http://git.xxxxxxxx.com/Echo.BSSR/MyProject; name=MyProject; name_with_namespace=Echo.BSSR / Echo.BSSR.Core.TLRGLoadT
enderOut; path=MyProject; path_with_namespace=Echo.BSSR/MyProject; resolve_outdated_diff_discussions=False; container_
registry_enabled=True; issues_enabled=True; merge_requests_enabled=True; wiki_enabled=True; builds_enabled=True; snippets_enabled=True; created_at=2018-05-30T21:43:
16.146Z; last_activity_at=2018-07-30T14:21:08.601Z; shared_runners_enabled=False; lfs_enabled=True; creator_id=204; namespace=; avatar_url=; star_count=0; forks_cou
nt=0; open_issues_count=0; public_builds=True; shared_with_groups=System.Object[]; only_allow_merge_if_build_succeeds=False; request_access_enabled=False; only_allo
w_merge_if_all_discussions_are_resolved=False; permissions=}.id  (try to get all branches on prior project) 
WARNING: No gitlab api token is defined for ECHOGL\someuser on computer VM3-someuser. Fix with add-GitLabToken
You cannot call a method on a null-valued expression.
At C:\Program Files (x86)\WindowsPowerShell\Modules\GitLab-API\obj\gitlabconnect.ps1:88 char:5
+     $header = @{
+     ~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At C:\Program Files (x86)\WindowsPowerShell\Modules\GitLab-API\Get-GitLabProjectBranch.ps1:58 char:3
+   $GitlabConnect.callapi($apiurl,$httpmethod,$parameters)
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Completed at 09/21/2018 16:26:34

```
Output of second run when I don't close Powershell ISE: 


```
Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      14409  1005    
Save-GitLabAPIConfiguration 
WARNING: All tokens will be sent over HTTP. Recommendation: Use HTTPS.
 
=Get-GitLabProject -Search 'MyProject'
You cannot call a method on a null-valued expression.
At C:\Program Files (x86)\WindowsPowerShell\Modules\GitLab-API\obj\gitlabconnect.ps1:88 char:5
+     $header = @{
+     ~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At C:\Program Files (x86)\WindowsPowerShell\Modules\GitLab-API\Get-GitLabProject.ps1:126 char:5
+     $GitlabConnect.callapi($apiurl,$httpmethod,$parameters)
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Get-GitLabProjectBranch -ProjectID .id  (try to get all branches on prior project) 
You cannot call a method on a null-valued expression.
At C:\Program Files (x86)\WindowsPowerShell\Modules\GitLab-API\obj\gitlabconnect.ps1:88 char:5
+     $header = @{
+     ~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At C:\Program Files (x86)\WindowsPowerShell\Modules\GitLab-API\Get-GitLabProjectBranch.ps1:58 char:3
+   $GitlabConnect.callapi($apiurl,$httpmethod,$parameters)
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
Completed at 09/21/2018 16:29:44

```

**Comment by ngetchell** [comment]
Hello @NealWalters 

All of the errors you mention are generated from the module GitLab-API, not PSGitLab. You have two options as it seems the two different modules use the same names. You can uninstall one or you can refer to each by the commands full name in the <ModuleName>\<Function> syntax. 

I've converted the above code snippet into the verbose syntax:

``` ps
PSGitLab\Save-GitLabAPIConfiguration
PSGitLab\Get-GitLabProject -Search 'MyProject'
PSGitLab\Get-GitLabProjectBranch -ProjectID $ID
```

This will force the use of PSGitLab over the GitLab-API. 

I hope this helps. Let me know if you have further questions!

**Comment by NealWalters** [comment]
I'm finally back... deleted and cleaned-up the old versions. 
Now it looks like Get-GitLabProjectBranch was deprecated? Doesn't exist in the current module when I do: 
get-command -Module PSGitLab

So how do I navigate to a specific branch? 

Thanks,
Neal

**Comment by ngetchell** [comment]
How do you mean navigate? You should be using the following locally to switch branches. 

``` bash
git checkout featureBranch
```
